_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+---
+layout: post
+title:  "Clear Separation of Specification and Implementation in Dafny"
+date:   2023-08-14 00:00:00 +0100
+author: Aaron Tomb
+categories:
+---
+
+Dafny allows you to describe computations in several ways,
+primarily pure functions and imperative methods, each with its
+own tradeoffs. Because of these tradeoffs, it can be useful to
+combine mechanisms and relate them to each other. This post
+describes one particularly effective technique for doing so, by
+clearly separating a _specification_ from an _implementation_,
+proving properties of the specification, and proving equivalence (or
+refinement) between the specification and the implementation.
+
+To get started, let's go over the tools Dafny provides. The two most
+common computational descriptions in Dafny are terminating,
+deterministic functions and imperative, stateful methods.
+
+By writing a `function`, you can describe a mathematical object that
+maps each input deterministically to a single output. Functions are
+typically concise, and, given their genesis as one of the
+foundational concepts in mathematics, they tend to be the most
+straightforward to reason about. Because of this, the bodies of
+functions are made directly available to the verifier by default,
+unless a function is marked `{:opaque}`. Functions can often also be
+compiled to executable programs, though sometimes inefficiently (and
+not at all if they make use of infinite data types).
+
+By writing a `method`, you can describe an imperative computation that
+performs a sequence of steps to yield an output state or a set of output
+states from each initial program state. Methods more closely match the
+operations performed by actual computing hardware, and are capable of
+describing non-deterministic or non-terminating computations. They are
+also generally meant to be compiled to executable programs. However,
+reasoning about methods is somewhat different than reasoning about
+functions. You can prove that a method agrees with a single contract,
+written as `requires`, `ensures`, `modifies`, and `decreases` clauses,
+but you generally do not expand the body of a method as part of
+performing a separate proof (though it is sometimes possible to do so,
```

Comment:
Remove parentheses to increase legibility. Blog posts can also use footnotes if necessary.